### PR TITLE
[TE] endpoint - harleyjj/yamlresource - add dataset as an optional pa…

### PR DESCRIPTION
**Pull request by harleyjj**
…ram for yaml/list

1) Makes dataset and metrics optional params for returning a filtered list of detection yaml configurations
2) Adds some Swagger annotations for yaml/list endpoint

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 1]
```diff
@@ -907,8 +907,28 @@ public Response toggleActivation(
...
+          .parallelStream()
+          .map(this.detectionConfigFormatter::format)
+          .collect(Collectors.toList());
+    } else {
+      yamls = this.detectionConfigDAO
+          .findAll()
+          .parallelStream()
+          .map(this.detectionConfigFormatter::format)
+          .filter(y -> {
+            List<String> thisYaml = (List)y.get("datasetNames");
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: variable "thisYaml" is not needed in this case.

Comment (harleyjj) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: I added metric as an optional param, as well.  Would this comment still apply to the updated code?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 2]
```diff
@@ -907,8 +907,46 @@ public Response toggleActivation(
...
+            }
+            if (dataset != null && datasetList != null) {
+              if (datasetList != null) {
+                datasetMatch = datasetList.contains(dataset);
+              } else {
+                datasetMatch = false;
+              }
+            }
+            return (metricMatch && datasetMatch);
+          })
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: The readability is bad. Let's rewrite this logic.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 3]
```diff
@@ -907,8 +907,46 @@ public Response toggleActivation(
...
+          .parallelStream()
+          .map(this.detectionConfigFormatter::format)
+          .collect(Collectors.toList());
+    } else {
+      yamls = this.detectionConfigDAO
+          .findAll()
+          .parallelStream()
+          .map(this.detectionConfigFormatter::format)
+          .filter(y -> {
+            List datasetList = (List) y.get("datasetNames");
```
[the remark this answers is not in the document]
Comment (harleyjj) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Method added.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java (hunk 2 above)
Comment (harleyjj) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: I rewrote the logic in the method

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 4]
```diff
@@ -900,15 +900,64 @@ public Response toggleActivation(
...
+      // equals method should not be called on null
+      metricMatch = (metricString != null) && metricString.equals(metric);
+    }
+    // check dataset only if it was passed
+    if (dataset != null) {
+      // contains method should not be called on null
+      datasetMatch = (datasetList != null) && datasetList.contains(dataset);
+    }
+    // config should satisfy both filters
+    return (metricMatch && datasetMatch);
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: return metricMatch && datasetMatch;

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 5]
```diff
@@ -900,15 +900,64 @@ public Response toggleActivation(
...
+    boolean datasetMatch = true;
+    // check metric only if it was passed
+    if (metric != null ) {
+      // equals method should not be called on null
+      metricMatch = (metricString != null) && metricString.equals(metric);
+    }
+    // check dataset only if it was passed
+    if (dataset != null) {
+      // contains method should not be called on null
+      datasetMatch = (datasetList != null) && datasetList.contains(dataset);
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: datasetMatch = datasetList != null && datasetList.contains(dataset);

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 6]
```diff
@@ -900,15 +900,64 @@ public Response toggleActivation(
...
+  private boolean filterConfigsBy(Map<String, Object> detectionConfig, String dataset, String metric) {
+    List datasetList = (List) detectionConfig.get("datasetNames");
+    String metricString = (String) detectionConfig.get("metric");
+    // defaults are true so we filter only if the params are passed
+    boolean metricMatch = true;
+    boolean datasetMatch = true;
+    // check metric only if it was passed
+    if (metric != null ) {
+      // equals method should not be called on null
+      metricMatch = (metricString != null) && metricString.equals(metric);
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: metricMatch = metricString != null && metricString.equals(metric);

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java [hunk 7]
```diff
@@ -904,15 +904,64 @@ public Response toggleActivation(
...
+    } else {
+      yamls = this.detectionConfigDAO
+          .findAll()
+          .parallelStream()
+          .map(this.detectionConfigFormatter::format)
+          .filter(y -> filterConfigsBy(y, dataset, metric))
+          .collect(Collectors.toList());
+    }
+    return yamls;
   }
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: Although the legacy code doesn't follow the standards, I have some suggestions which we have been trying to follow for all the new apis.
1. If there is an exception thrown during the execution then we need to wrap it up and return a meaningful response to the caller along with the status codes.
2. Also consider adding some logging statements indicating how many configurations were returned for the params passed.

Refer example: https://github.com/apache/incubator-pinot/blob/master/thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/api/user/dashboard/UserDashboardResource.java#L220

Comment (harleyjj) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/yaml/YamlResource.java]: I made these changes.  Please take a look.

**Comment by xiaohui-sun** [comment]
One request not specific to this PR: Can we show error message in a separate panel? Currently it only shows as pop up window which is not very readable. @harleyjj

**Pull request merged**